charms/jupyter-ui/src/charm.py

```diff
@@ -8,63 +8,67 @@
 from ops.charm import CharmBase
 from ops.main import main
 from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus, WaitingStatus
-from serialized_data_interface import NoCompatibleVersions, NoVersionsListed, get_interfaces
+from serialized_data_interface import (
+    NoCompatibleVersions,
+    NoVersionsListed,
+    get_interfaces,
+)
 
 log = logging.getLogger()
 
+class CheckFailed(Exception):
+    """ Raise this exception if one of the checks in main fails. """
+
+    def __init__(self, msg, status=None):
+        super().__init__()
+
+        self.msg = msg
+        self.status = status
 
 class Operator(CharmBase):
     def __init__(self, *args):
         super().__init__(*args)
 
-        if not self.model.unit.is_leader():
-            log.info("Not a leader, skipping set_pod_spec")
-            self.model.unit.status = WaitingStatus("Waiting for leadership")
-            return
-
-        try:
-            self.interfaces = get_interfaces(self)
-        except NoVersionsListed as err:
-            self.model.unit.status = WaitingStatus(str(err))
-            return
-        except NoCompatibleVersions as err:
-            self.model.unit.status = BlockedStatus(str(err))
-            return
-        else:
-            self.model.unit.status = ActiveStatus()
-
         self.image = OCIImageResource(self, "oci-image")
+        for event in [
+            self.on.install,
+            self.on.leader_elected,
+            self.on.upgrade_charm,
+            self.on.config_changed,
+        ]:
+            self.framework.observe(event, self.main)
+            # FIXME: I am unsure of this approach, old charm called send_info()
+            # for every event, and that is why I left it. We can remove
+            # it if it is no longer necessary
+            self.framework.observe(event, self.configure_mesh)
 
-        self.framework.observe(self.on.install, self.set_pod_spec)
-        self.framework.observe(self.on.upgrade_charm, self.set_pod_spec)
-        self.framework.observe(self.on.config_changed, self.set_pod_spec)
-
-        self.framework.observe(self.on.install, self.send_info)
-        self.framework.observe(self.on.upgrade_charm, self.send_info)
-        self.framework.observe(self.on.config_changed, self.send_info)
-        self.framework.observe(self.on['ingress'].relation_changed, self.send_info)
+        self.framework.observe(self.on['ingress'].relation_changed, self.configure_mesh)
```

Comment:
You will also have to call main with the `ingress` relation_changed event and call `configure_mesh` from there. See this [PR](https://github.com/canonical/kubeflow-dashboard-operator/pull/12/files) for reference.

Additionally all your tests are failing because `self.interfaces` is not initialized anymore, you have to passthrough `interfaces` to `configure_mesh`

Reply:
The thing is (and this is related to the comment in L40), the old charm calls `self.set_pod_spec` and `self.send_info` for all (`install`, `upgrade_charm`, and `config_changed`) except for `relation_changed` as it ONLY calls `self.send_info`, and that's why I left `self.framework.observe(self.on['ingress'].relation_changed, self.configure_mesh)` outside of the event list. If there is no problem with that event calling `main()`, I can change it, I just want to make sure this is the behaviour we want.

Reply:
Nvm, I just checked the latest changes. Will update this. Thanks.